lib/scripts.js

```diff
@@ -371,6 +371,68 @@ var scripts = {
 
     return execScript.apply(scripts, args);
   },
+
+  cleanJobsInSet: function(queue, set, ts) {
+    var command;
+    var hash;
+    switch(set) {
+      case 'completed':
+      case 'failed':
+        command = 'local JOBS = redis.call("SMEMBERS", KEYS[1])';
+        hash = 'cleanSet';
+        break;
+      case 'wait':
+      case 'active':
+      case 'paused':
+        command = 'local JOBS = redis.call("LRANGE", KEYS[1], 0, -1)';
+        hash = 'cleanList';
+        break;
+      case 'delayed':
+        command = 'local JOBS = redis.call("ZRANGE", KEYS[1], 0, -1)';
+        hash = 'cleanOSet';
+        break;
+    }
+
+    var script = [
+      command,
+      'local deleted = {}',
+      'local jobTS',
+      'for _, job in ipairs(JOBS) do',
+      '  if (redis.call("EXISTS", ARGV[1] .. job ..  ":lock") == 0) then',
+      '    jobTS = redis.call("HGET", ARGV[1] .. job, "timestamp")',
+      '    if(not jobTS or jobTS < ARGV[2]) then',
+      '      redis.call("LREM", ARGV[1] .. ARGV[3], 0, job)',
+      '      redis.call("LREM", ARGV[1] .. ARGV[4], 0, job)',
+      '      redis.call("ZREM", ARGV[1] .. ARGV[5], job)',
+      '      redis.call("LREM", ARGV[1] .. ARGV[6], 0, job)',
+      '      redis.call("SREM", ARGV[1] .. ARGV[7], job)',
+      '      redis.call("SREM", ARGV[1] .. ARGV[8], job)',
+      '      redis.call("DEL", ARGV[1] .. job)',
```

Comment:
I actually attempted to do this, I saw a 10% gain on running time, which is not bad, but I wanted to publish the PR first to get some feedback first.

Originally `Queue#clean` uses `Job#remove` to clean the jobs, which in turns removes from all sets, so I followed that logic here. However, I agree that a job must exist in one place only.

If we have a lock, we can guarantee that a given job is in only one set and will remain so for all different states, so we can make improvements there.

The above reminds me of: https://github.com/OptimalBits/bull/pull/291#issuecomment-217281186 I now understand why there was that check in the first place, `LREM` is a O(n) operation, so it was issued only if needed for performance reasons, however the only way we could guarantee that this is the case is if we have a job lock, otherwise it would leave jobs in a corrupted state in some scenarios (I noticed that there were jobs in `completed` but that had no job data). Since we are taking locks on removal, we could probably bring back that check to improve `Job#remove`.